FastText/src/models/custom_layers.jl

```diff
@@ -0,0 +1,267 @@
+    maskWh = Flux.dropout_mask(Flux.rng_from_array(), cell.Wh, cell.p)
+    hidden = (cell.state0..., maskWi, maskWh)
+    return Flux.Recur(cell, hidden)
+end
+
+function Flux.reset!(layer::Flux.Recur{<:WeightDroppedLSTMCell})
+    maskWi = Flux.dropout_mask(Flux.rng_from_array(), layer.cell.Wi, layer.cell.p)
+    maskWh = Flux.dropout_mask(Flux.rng_from_array(), layer.cell.Wh, layer.cell.p)
+    layer.state = (layer.cell.state0..., maskWi, maskWh)
+    return nothing
+end
+####################################################################
+
+########################## Varitional DropOut ######################
+"""
+    VarDrop(p::Float64=0.0)
+
+Variational Dropout layer
+
+Unlike standard dropout layer, which applies new dropout mask to every Array ot Matrix passed to it,
+this layer saves the mask applied till it is explicitly set to reset mode using 'reset_masks!'.
+
+Usage:
+julia> vd = VarDrop()
+
+To reset mask:
+julia> reset_masks!(vd)
+"""
+
+mutable struct VarDropCell{F}
+    p::F
+    active::Union{Bool,Nothing} # matches other norm layers
+end
+Flux.@functor VarDropCell
+
+VarDropCell(p::Real = 0.0) = VarDropCell(p, nothing)
+
+testmode!(m::VarDropCell, mode = true) =
+    (m.active = (isnothing(mode) || mode == :auto) ? nothing : !mode; m)
+
+
+function (vd::VarDropCell)((has_mask, mask), x)
+    if Flux._isactive(vd)
+        mask = has_mask ? mask : Flux.dropout_mask(Flux.rng_from_array(), x, vd.p)
+        return (true, mask), x .* mask
+    elseif !has_mask
+        return (has_mask, mask), x
+    else
+        error("Mask set but layer is in test mode. Call `reset!` to clear the mask.")
+    end
+end
+
+# The single-element array keeps Recur happy.
+# Limitation: typeof(p) must == typeof(<inputs>)
+VarDrop(p::Real) = Flux.Recur(VarDropCell(p), (false, ones(typeof(p), 1, 1)))
+
+function Flux.reset!(m::Flux.Recur{<:VarDropCell})
+    m.state = (false, ones(typeof(m.cell.p), 1, 1))
+    return nothing
+end
+
+
+######################################################################
+
+################# Varitional Dropped Embeddings ######################
+"""
+    DroppedEmbeddings(in::Integer, embed_size::Integer, p::Float64=0.0)
+
+Embeddings with varitional dropout
+
+This struct defines an embedding layer with Varitional Embedding dropout functionality.
+Instead of randomly dropping values of embedding matrix,
+this layer drops all values of a specific token, in other words,
+that token is dropped from the embedding matrix for that particular pass.
+
+Since this follows Variational DropOut criteria, it also saves the drop mask,
+which should be reset to new mask explicilty using `reset_masks!` function
+
+# Usage:
+It takes input size, embedding size and dropout probability
+
+julia> de = DroppedEmbeddings(1000, 20, 0.4)
+
+To reset mask:
+
+julia> reset_masks!(de)
+"""
+mutable struct DroppedEmbeddings{A,F}
+    emb::A
+    p::F
+    mask::Vector{Float32}
```

Comment:
This has to be an abstract array type so that GPU arrays (`CuVector`s) also work